File: atcoder/abc265D_iroha_and_haiku.hs
-- uninhm
-- https://atcoder.jp/contests/abc265/tasks/abc265_d
-- binary search, accumulation

import Data.Array.IArray
import Data.Maybe (isJust)
import Data.Bool (bool)

binarySearch :: Ord a => Array Int a -> a -> Maybe Int
binarySearch a x = go 0 (length a) where
  go l r
    | elem == x = Just l
    | l == r    = Nothing
    | elem < x  = go (mid+1) r
    | otherwise = go l mid
    where mid = (l + r) `div` 2
          elem = a ! mid

getInts :: IO [Int]
getInts = map read . words <$> getLine

accumulate :: [Int] -> [Int]
accumulate l = reverse $ go l [0] where
  go [] res = res
  go (x:l) (y:res) = go l $ (x+y):y:res

solve :: (Int, Int, Int) -> Array Int Int -> Int -> Maybe ()
solve (p, q, r) a w = do
  binarySearch a (a!w - r)
  binarySearch a (a!w - r - q)
  binarySearch a (a!w - r - q - p)
  pure ()

main :: IO ()
main = do
  [n, p, q, r] <- getInts
  a <- listArray (0,n) . accumulate <$> getInts
  putStrLn $ bool "No" "Yes" $ any isJust $ map (solve (p, q, r) a) [3..n]

File: atcoder/abc265D_iroha_and_haiku.py
# uninhm
# https://atcoder.jp/contests/abc265/tasks/abc265_d
# binary search, cumulative sum

# The idea:
# Once you have the w, you can binary search your way back to x
# Testing every possible w should be O(n*log n)

def binary_search(a, x):
    l = 0
    r = len(a)
    while l < r:
        mid = (l+r)//2
        if a[mid] < x:
            l = mid + 1
        else: # a[mid] >= x
            r = mid
    return l

n, p, q, r = map(int, input().split())
a = list(map(int, input().split()))
acc = [0]
for x in a:
    acc.append(acc[-1]+x)

for w in range(n, 2, -1):
    z = binary_search(acc, acc[w]-r)
    y = binary_search(acc, acc[z]-q)
    x = binary_search(acc, acc[y]-p)
    if (p, q, r) == (acc[y]-acc[x], acc[z]-acc[y], acc[w]-acc[z]):
        print("Yes")
        break
else:
    print("No")
